File: scripts/lexical/pt/sidecar.mjs
/**
 * Node wrapper for `sidecar.py` (simplemma): batches sentences, shells out
 * once per batch, caches per-input results under
 * `artifacts/lexical/pt/<sha1>.json` (gitignored). Mirrors
 * `scripts/lexical/es/sidecar.mjs`; see that file (and `ja/sidecar.mjs`)
 * for the caching rationale. Kept as its own copy, not a shared import,
 * matching the JA/ES/FR sidecar's per-language-directory convention.
 */
import { spawnSync } from "node:child_process";
import { createHash } from "node:crypto";
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const LANG = "pt";
const HERE = path.dirname(fileURLToPath(import.meta.url));
const REPO_ROOT = path.resolve(HERE, "../../../");
const CACHE_DIR = path.join(REPO_ROOT, `artifacts/lexical/${LANG}`);
// Env-overridable (2026-09-17, lane A7f) — see scripts/lexical/ja/sidecar.mjs's
// matching comment for the rationale.
const PYTHON = process.env.LINGO_LEXICAL_PYTHON_PT || process.env.LINGO_LEXICAL_PYTHON || path.join(HERE, ".venv/bin/python");
const SIDECAR_PY = path.join(HERE, "sidecar.py");

function sha1(s) {
  return createHash("sha1").update(s, "utf8").digest("hex");
}

function cachePath(text) {
  return path.join(CACHE_DIR, `${sha1(text)}.json`);
}

function readCache(text) {
  const p = cachePath(text);
  if (!existsSync(p)) return null;
  try {
    return JSON.parse(readFileSync(p, "utf8"));
  } catch {
    return null;
  }
}

function writeCache(text, tokens) {
  mkdirSync(CACHE_DIR, { recursive: true });
  writeFileSync(cachePath(text), JSON.stringify(tokens), "utf8");
}

let venvChecked = false;
export function sidecarAvailable() {
  if (!venvChecked) {
    venvChecked = existsSync(PYTHON) && existsSync(SIDECAR_PY);
  }
  return venvChecked;
}

export function tagBatch(items) {
  const byText = new Map();
  for (const { id, text } of items) {
    if (!byText.has(text)) byText.set(text, []);
    byText.get(text).push(id);
  }

  const toRun = [];
  const resolved = new Map();
  for (const text of byText.keys()) {
    const cached = readCache(text);
    if (cached) resolved.set(text, cached);
    else toRun.push(text);
  }

  if (toRun.length > 0) {
    if (!sidecarAvailable()) {
      throw new Error(
        `${LANG.toUpperCase()} lexical sidecar not installed. Run:\n  cd scripts/lexical/${LANG} && uv venv .venv --python 3.11 && uv pip install --python .venv/bin/python -r ../requirements-pt.txt`,
      );
    }
    const payload = toRun.map((text, i) => ({ id: String(i), text }));
    const proc = spawnSync(PYTHON, [SIDECAR_PY, LANG], {
      input: JSON.stringify(payload),
      encoding: "utf8",
      maxBuffer: 64 * 1024 * 1024,
    });
    if (proc.status !== 0) {
      throw new Error(`sidecar.py failed (exit ${proc.status}): ${proc.stderr}`);
    }
    const out = JSON.parse(proc.stdout);
    for (let i = 0; i < toRun.length; i++) {
      const text = toRun[i];
      const tokens = out[i].tokens;
      resolved.set(text, tokens);
      writeCache(text, tokens);
    }
  }

  const result = new Map();
  for (const [text, ids] of byText) {
    const tokens = resolved.get(text) ?? [];
    for (const id of ids) result.set(id, tokens);
  }
  return result;
}

export function tagOne(text) {
  return tagBatch([{ id: "0", text }]).get("0");
}

File: scripts/lexical/pt/sidecar.py
#!/usr/bin/env python3
"""
PT lexical sidecar: per-tile/per-sentence tokens (surface, lemma, is_known)
via simplemma, for the procedural-QA runner's PT port
(`docs/pt-course-design-2026-09-18.md` §3, "Lexical sidecar for PT
procedural QA").

Install (project venv, exact command — record any drift here):
    cd scripts/lexical/pt
    uv venv .venv --python 3.11
    uv pip install --python .venv/bin/python -r ../requirements-pt.txt

Licence: simplemma is MIT (Adrien Barbaresi, 2021 — see
`.venv/lib/python3.11/site-packages/simplemma-*.dist-info/licenses/LICENSE`),
~19MB installed (lemmatization dictionaries bundled as package data — no
separate download step, unlike JMdict/Lexique).

simplemma has no morphological-analyzer POS tagger (unlike fugashi/Kiwi) —
it offers `lemmatize` (dictionary + rule-based lemma lookup) and
`is_known` (is this surface, or its lemma, in the dictionary at all). For
PT this sidecar is used as a LEMMA + DICTIONARY-MEMBERSHIP fact source,
not a POS source — a course-specific compositional word list (course
atoms, `authoring-invariants-pinned.md`-equivalent structural facts) is
still the primary check, this is the fallback the way fugashi is the JA
Q3 fallback.

## Usage

    echo '[{"id":"a","text":"Eu sou estudante"}]' | .venv/bin/python sidecar.py pt

Reads a JSON array of {id, text} from stdin, a two-letter simplemma lang
code as argv[1] (pt — simplemma's `pt` language code; this file is kept as
its own literal copy per language like the JA/ES sidecar convention, not a
shared import, so each language's venv is independently recreatable),
writes a JSON array of {id, tokens: [{surface, lemma, isKnown}]} to
stdout — one token per `simple_tokenizer` piece (whitespace + punctuation
split; PT is space-tokenized, so this is closer to a naive word split
than fugashi's morphological segmentation, which is the whole reason
Q2/Q3 for PT are about MULTI-WORD TILES, not sub-word morphemes — see the
doc's "chunk" definition).
"""
import json
import sys

try:
    import simplemma
except ImportError as e:  # pragma: no cover - operational failure, not a test path
    sys.stderr.write(
        "simplemma is not installed. Run:\n"
        "  cd scripts/lexical/pt && uv venv .venv --python 3.11 && "
        "uv pip install --python .venv/bin/python -r ../requirements-pt.txt\n"
    )
    raise


def tag_one(text, lang):
    tokens = []
    for surface in simplemma.simple_tokenizer(text):
        if not surface.strip():
            continue
        lemma = simplemma.lemmatize(surface, lang=lang)
        known = simplemma.is_known(surface, lang=lang) or simplemma.is_known(
            surface.lower(), lang=lang
        )
        tokens.append({"surface": surface, "lemma": lemma, "isKnown": bool(known)})
    return tokens


def main():
    lang = sys.argv[1] if len(sys.argv) > 1 else "pt"
    raw = sys.stdin.read()
    items = json.loads(raw) if raw.strip() else []
    out = []
    for item in items:
        out.append({"id": item["id"], "tokens": tag_one(item["text"], lang)})
    json.dump(out, sys.stdout, ensure_ascii=False)


if __name__ == "__main__":
    main()
